security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,375 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package advancedtls is a utility library containing functions to construct
+// credentials.TransportCredentials that can perform credential reloading and custom
+// server authorization.
+// All APIs in this package are EXPERIMENTAL.
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"syscall"
+	"time"
+
+	"google.golang.org/grpc/credentials"
+)
+
+type sysConn = syscall.Conn
+
+// CustomVerificationFunc is the function defined by users to perform custom server
+// authorization.
+type CustomVerificationFunc func(serverName string,
+	rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
+
+// NewClientTLS uses ClientOptions to construct a TransportCredentials based on TLS.
+func NewClientTLS(o *ClientOptions) (credentials.TransportCredentials, error) {
+	conf, err := o.config()
+	if err != nil {
+		return nil, err
+	}
+	tc := &advancedTLSCreds{
+		config:     conf,
+		isClient:   true,
+		getRootCAs: o.GetRootCAs,
+		verifyFunc: o.VerifyPeer,
+	}
+	tc.config.NextProtos = appendH2ToNextProtos(tc.config.NextProtos)
+	return tc, nil
+}
+
+// NewServerTLS uses ServerOptions to construct a TransportCredentials based on TLS.
+func NewServerTLS(o *ServerOptions) (credentials.TransportCredentials, error) {
+	conf, err := o.config()
+	if err != nil {
+		return nil, err
+	}
+	tc := &advancedTLSCreds{
+		config:     conf,
+		isClient:   false,
+		getRootCAs: o.GetRootCAs,
+	}
+	tc.config.NextProtos = appendH2ToNextProtos(tc.config.NextProtos)
+	return tc, nil
+}
+
+// ClientOptions contains all the fields and functions needed to be filled by the client.
+type ClientOptions struct {
```

Comment:
I've thought of some ways of extraction, but the input parameters of GetClientCertificate(for client) and GetCertificate(for server) are different. It might look weird if we extract the other 3 and call them "CertificateOptions" ... Do we have any better ways for this?

Reply:
I was thinking of something like the following:

```
type CertificateOptions struct {
    // Options common for clients and servers.
}

type ClientOptions struct {
    CertOptions CertificateOptions
    // The rest of client options.
}

type ServerOptions struct {
    CertOptions CertificateOptions
    // The rest of server options.
}
```